modules/core/include/opencv2/core/hal/intrin_neon.hpp

```diff
@@ -726,41 +726,61 @@ inline v_float64x2 v_dotprod_expand(const v_int32x4& a,   const v_int32x4& b,
 // 16 >> 32
 inline v_int32x4 v_dotprod_fast(const v_int16x8& a, const v_int16x8& b)
 {
+#if CV_SIMD128_64F
+    int32x4_t p = vmull_s16(vget_low_s16(a.val), vget_low_s16(b.val));
+    return v_int32x4(vmlal_high_s16(p, a.val, b.val));
```

Comment:
> CV_SIMD128_64F

In general, `64F` should be used with the `double` type.

Why does `#ifdef __aarch64__` not work here? (or other condition)

Reply:
It's misleading, but `vmlal_high_s16` is only available for `aarch64` [ref](https://developer.arm.com/architectures/instruction-sets/simd-isas/neon/intrinsics?search=vmlal_high_s16)
The macro might be misleading, but it's doing the right thing.

Reply:
My proposal is:
```
-#if CV_SIMD128_64F
+#ifdef __aarch64__
```

Reply:
Hello @alalek , @tomoaki0705  - thank you for the review!

My initial downstream code was using `__aarch64__`, but than I thought to use `CV_SIMD128_64F` to catch both clang and gcc, and also visual studio:

```
if defined(__aarch64__) || defined(_M_ARM64)
#define CV_SIMD128_64F 1
#else
#define CV_SIMD128_64F 0
#endif
```
`_M_ARM64` is equivalent to `__aarch64__` when targeting Windows.

I agree the name `CV_SIMD128_64F` is misleading in this case. We are not conditionally compiling for checking if there is support for `double` types, we are checking if we are in `aarch64` execution mode on `armv8`.

I would like to capture as many compiler as possible with one flag, so I would like to propose the following macro to be used for the code in this patch.

```
if defined(__aarch64__) || defined(_M_ARM64)
#define CV_NEON_ADVSIMD 1
#else
#define CV_NEON_ADVSIMD 0
#endif
```

The name `Advanced SIMD` is the name of NEON on Armv8: https://developer.arm.com/documentation/101028/0012/13--Advanced-SIMD--Neon--intrinsics?lang=en

Would that be OK?

Reply:
Right. My review was missing the view of Visual Studio.
In that case, I vote for re-using `CV_SIMD128_64F`
`CV_SIMD128_64F` is used in many places, and I don't like renaming it as `CV_NEON_ADVSIMD` (Though, I like the idea).
My suggestion is to use just `CV_SIMD128_64F` and remove `__aarch64__`.  That'll make things more simple.

Reply:
Sure - thanks. 

BTW, I didn't mean to rename the old one, I meant to add a new one. :)

But leaving just one macro is easier and simpler!

So I will leave it as it is.

Thank you again!

I'll push the other fixes in a minute.

Reply:
Well, please wait the comment from @alalek or someone else who can merge this.
I'm not authorized for merging

Reply:
Of course, @tomoaki0705 . Thank you again for your review. :)

Reply:
`CV_SIMD128_64F` macro is a part of external SIMD API, so we should not change its meaning.

What is about `CV_NEON_AARCH64` new macro (just because of MSVS)? (it is more clear than ADVSIMD)

Anyway it makes sense to add comment near this definition with corresponding link on ARM documentation.

Reply:
`CV_NEON_AARCH64` is fine. I'll update the patch and use that. I'll add the comment to the documentation.